### PR TITLE
Code refactor part 1

### PR DESCRIPTION
**Pull request by yakiza**
### **Description**

After carefully reading the documentation and going through the code, considering the 7 improvement ideas, I have chosen to propose an alternative. My suggestion is to refactor the existing code and not add anything more on top. This will help in the future development and maintainability of the code for both existing developers and newcomers. The goal is to decouple the code by using interfaces and more suitable packages, making it more idiomatic and modular. This will result in a codebase that is more in line with Go's style, and it will be extendable and maintainable.

Since this is going to be a substantial pull request, I have decided to make it more manageable. To achieve this, I have chosen to include the logic for each component in separate commits. This way, each commit can be reviewed independently, following the progression of the changes.

### First Commit
First, we will relocate all the code from the `cmd/` package to their respective appropriate packages. In Go projects, it is a common convention to use the `cmd/` package specifically for executable entry points and its usually  the place to wire up the various services and fire up server. By following this convention, we can achieve the following benefits 

- **Improved organization:** Placing only the executable-related code in the `cmd/` directory provides a clear separation between the application's core logic and the entry points. It helps developers quickly identify and navigate to the entry points, making the project structure more intuitive and maintainable.
- **Ease of distribution:** When we isolate the executables in the `cmd/` directory, it simplifies the process of building and distributing the application. We can easily compile and package the binaries independently, making it straightforward to distribute them as standalone units or deploy them to different environments.
- **Support for multiple binaries:** With the `cmd/` directory dedicated to entry points, it becomes easier to add and manage multiple binaries within the same project. Each binary can have its own separate configuration, dependencies, and build settings, allowing for flexibility in supporting various application modes or deployment scenarios.
- **Go idiomatic structure:** Placing the executable code in the `cmd/` directory follows the idiomatic Go project structure, which is widely recognized and adopted by the Go community. By adhering to established conventions, the codebase becomes more familiar and easier to understand for experienced Go developers. It reduces the cognitive load when navigating through the project, as developers already know where to find the entry points and can quickly identify the primary executables.

### Second Commit 
 
Our objective with the code refactor is to achieve a clear separation of responsibilities. To accomplish this, we will introduce an API layer. This will involve creating an `api/` package that will house all the code responsible for communication-related tasks.

By implementing this API layer, we ensure that the code responsible for handling communication is contained in a dedicated package. This separation helps improve modularity and maintainability. The` api/` package will handle tasks such as routing, request handling, and response generation, allowing for a more organized and focused codebase.

The introduction of an API layer brings several advantages, including better encapsulation, ease of testing, and improved code readability. It enables developers to understand and work with the communication aspects of the codebase more effectively.

For the API, I have chosen to include go-chi (https://github.com/go-chi/chi), which is an idiomatic lightweight router for Go HTTP services. It allows us to handle routing more efficiently and cleanly, and also provides several useful middlewares, such as the Logger and Recoverer middleware that we have introduced. In the initial stage, we have maintained the grouping of handlers as it was in the code. This means that we have api/bridge_handler.go, which is responsible for the bridge connection, and api/ui_handler, which is responsible for all the activities coming from the UI. In the next iteration, this will be further improved by creating api/session_handlers.go, and so on.

Furthermore, semantic erroring has been implemented to improve error handling. Semantic erroring refers to the practice of using descriptive error messages and codes that convey meaningful information about the nature and cause of the error.

- I**mproved Debugging**: Clear and informative error messages help developers quickly identify the source of an error. By providing specific details about the error, such as the module or function where it occurred, the specific inputs that caused the error, or the expected behavior, semantic error messages facilitate faster debugging and troubleshooting.
- **Effective Error Handling:** Semantic error codes enable more effective error handling strategies. By assigning unique error codes to different types of errors, developers can implement targeted error handling logic. This allows for fine-grained control over error handling, such as selectively catching specific errors, providing appropriate recovery mechanisms, or escalating critical errors to higher-level error handlers.
- **Maintainable and Extensible**: Semantic erroring promotes code maintainability and extensibility. By using consistent and descriptive error messages, future developers working on the codebase can quickly understand the purpose and context of each error. This makes it easier to maintain and extend the software over time, as developers can easily identify and handle different error scenarios.

We have also incorporated the internal package, which is a special package in Go that serves a specific purpose.
The internal package allows for encapsulation and information hiding within a Go project. By placing certain code or components inside the internal package, you limit their visibility to only the modules within the same project. This provides a clean and well-defined boundary between public and private functionality. External users or libraries cannot access or rely on internal components, reducing the risk of unintentional dependencies or misuse.

Note: Some functionalities, such as sync_stages, reorg, remote_db, and logs, have not yet undergone complete refactoring. I began noticing significant changes and will address them in the next iteration. However, I have made substantial progress in refactoring versions, flags, sessions, and cmd_args, bringing them closer to the desired end result.

### Third  & Forth Commit

Decoupling the responsibilities of the API layer from the business logic begins with sessions, as it is one of the first components we will be working with and require. To accomplish this, we have created a package called 'sessions/' that will contain the necessary logic. Furthermore, we have moved some of the methods from the 'ui_handler' to 'sessions' since they conceptually belong there. This approach allows us to maintain clean and focused handlers that deal with a single responsibility.

Additionally, you will notice that we have started incorporating interfaces into our codebase. Interfaces play a vital role in decoupling our code by providing them to the adjacent layers. By utilizing interfaces, developers can work on different layers of a feature concurrently, as they only need to work with the defined interface. Furthermore, if the implementation of one layer changes, it does not break the other layers that rely on it, as they are not tightly coupled to the specific implementation.

Additionally, we will introduce a new package called erigon_node, which will be responsible for managing node data. This package will include the "client" (previously known as remote_api).

In terms of naming conventions, we have made considerations to avoid unnecessary repetition in the code. For instance, within the sessions package, the add method does not include the term "session" in its name. Instead of using "AddSession", which would be redundant when called as sessions.Add(), we simply use sessions.Add()

Note: Some of the fields have been temporarily exported as part of an ongoing edit. These changes are subject to be included in the upcoming refactor.

### Fifth Commit ++

Directories that serve a generic purpose and are not mandatory for development, such as "images" and "demo_tls," now have an underscore (_) prefix. This naming convention helps us maintain a clean repository, particularly when working with an IDE. By placing these directories at the top of the project, they are conveniently grouped together.

Now that we have finished cleaning up the directories, let's proceed by adding a Makefile to simplify our lives when it comes to running, building, and testing our project, whether with or without Docker. Additionally, we will create a Dockerfile that can be utilized by the Makefile to create and spin up a container.

Lastly, we are introducing our semantic error handling files for "not found" and "bad request" scenarios. Although we currently have these files for only these two methods, we have plans to add more in the future, such as "StatusUnauthorized," "StatusMethodNotAllowed," "StatusForbidden," and so on

Extra nuggets:

- **Naming styles**: Inconsistent use of snake case, such as "current_sessionname," has been addressed and fixed to "current_session_name." Similarly, "session_name" has also been adjusted.
- **JavaScript variable type:** The code has been refactored to adhere to the recommendation of using "let" instead of "var" for variable declarations. This improves code readability and helps prevent unintended scoping issues.
- **Code deduplication:** Efforts have been made to eliminate code duplication, resulting in a more streamlined and efficient codebase. Duplicate code segments have been identified and refactored to enhance maintainability and reduce redundancy.

To enhance the data transfer process between JavaScript and Go, we should remove the use of the addForm method and instead send the data in a structured JSON format. Go is well-equipped to handle JSON data efficiently. By leveraging the built-in capabilities of Go's JSON handling, we can ensure a more seamless and standardized data exchange between the two languages. This approach can improve code clarity and maintainability while taking advantage of Go's robust JSON handling capabilities.

Have thoroughly tested the code to the best of my current knowledge, and it appears that everything is functioning as expected. I would appreciate any feedback or suggestions regarding testing edge cases to ensure that the code behaves consistently with its initial intentions.

If we decide to merge this in, README.md will be updated before it gets merged to reflect the new architecture schematics and details

### CODE REVIEW

File: api/helpers.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package api
+
+import (
+	"fmt"
+	"github.com/ledgerwatch/diagnostics"
+	"github.com/pkg/errors"
```
Comment (AskAlexSharov) [inline api/helpers.go:6]: do you see any advantage of using `pkg/errors` over native errors.Is/errors.As/fmt.Printf("%w") introduced in go 1.13? https://go.dev/blog/go1.13-errors

File: cmd/diagnostics/main.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	signalCh := make(chan os.Signal, 1)
+	signal.Notify(signalCh, syscall.SIGTERM, syscall.SIGINT)
+
+	// Initialize Services
+	cache := sessions.NewCache(5, 5)
+	ErigonNodeClient := erigon_node.NewErigonNodeClient()
+	uiSessions := sessions.NewUISession(cache)
+	htmlTemplates, err := template.ParseFS(assets.Templates, "template/*.html")
+	if err != nil {
+		fmt.Errorf("parsing session.html template: %v", err)
```
[the remark this answers is not in the document]
Comment (yakiza) [reply cmd/diagnostics/main.go:37]: absolutely, thank you, addressing.

File: assets/template/session.html [hunk 3]
```diff
@@ -6,11 +6,11 @@
     </head>
     <body>
         <form action="" method="post" id="session_form">
```
[the remark this answers is not in the document]
Comment (yakiza) [reply assets/template/session.html:8]: We can certainly improve it and have the "enter" work  as someone would expect base on the fields and the  "form or action that is being  completed" 

I would suggest that we improve the user experience and user interface design, we can split the form in 1 or two forms and have better handling, or we can keep it as is and enhance the functionality.

**Comment by AskAlexSharov** [comment]
@yakiza please give some advise for our .css (don't give any - is also ok).

**Comment by dntreou** [comment]
@AskAlexSharov  To be quite honest I would like us to separate our UI from our API.

I am assuming that this was a solution to a problem that needed addressing and works for now, since it delivers value in troubleshooting our Erigon nodes, but an adoption of a front end framework like angular can help in clearly separating concerns provide a better user experience and user interface. 

To answer your question I would like to use a CSS preprocessor like SASS, use Flexbox for layout, introduce media queries for a responsive UI , we could even use something like bootstrap I guess but also I would say we can use a BEM methodology.

**Pull request merged**